File: src/actions/ActionValidator.js
import {ActionParamSchema,ALWAYS_CONFIRM,isKnownAction} from './ActionSchema';
export function validateAction(a){
 if(!a||typeof a!=='object') return {valid:false,reason:'invalid_action'};
 if(!isKnownAction(a.action)) return {valid:false,reason:`unknown_action:${a.action}`};
 for(const k of ActionParamSchema[a.action]||[]) if(a.params?.[k]===undefined||a.params?.[k]===null) return {valid:false,reason:`missing_param:${k}`};
 if(ALWAYS_CONFIRM.has(a.action)&&a.requiresConfirmation!==true) return {valid:false,reason:`confirmation_required:${a.action}`};
 return {valid:true};
}
export function validatePlan(p){
 if(!p||p.type!=='ACTION_PLAN'||!Array.isArray(p.actions)) return {valid:false,reason:'invalid_plan'};
 if(p.actions.length>25) return {valid:false,reason:'too_many_actions'};
 for(const a of p.actions){const v=validateAction(a);if(!v.valid)return v}
 return {valid:true};
}

File: android/app/src/main/java/com/fixmyphone/executor/ActionValidator.kt
package com.fixmyphone.executor

/**
 * Native-side mirror of src/actions/ActionValidator.js. Runs the same check
 * a second time on the Kotlin side so that a bug (or tampering) in the JS
 * layer can never get an unvalidated or under-specified action past this
 * point — nothing reaches an ActionHandler without going through here.
 */
data class ValidationResult(
    val valid: Boolean,
    val reason: String? = null
)

object ActionValidator {

    fun validate(action: FixMyPhoneAction): ValidationResult {
        val requiredParams = ActionParamSchema.required[action.action]
            ?: return ValidationResult(false, "unknown_action:${action.action}")

        for (key in requiredParams) {
            if (!action.params.containsKey(key) || action.params[key] == null) {
                return ValidationResult(false, "missing_param:$key")
            }
        }

        // Belt-and-braces: even if the JS layer forgot to set
        // requiresConfirmation for an always-confirm action, we treat it as
        // requiring confirmation here. A caller that skips confirmation for
        // one of these gets rejected, not silently downgraded.
        if (action.action in ActionParamSchema.alwaysConfirm && !action.requiresConfirmation) {
            return ValidationResult(false, "confirmation_required:${action.action}")
        }

        return ValidationResult(true)
    }
}
